src/Marten/Services/UnitOfWork.cs

```diff
@@ -342,7 +336,73 @@ public void Eject<T>(T document)
             {
                 operations.Remove(operation);
             }
-           
+        }
+
+        private class StorageOperationComparer : IComparer<IStorageOperation>
+        {
+            private readonly Type[] _topologicallyOrderedTypes;
+
+            public StorageOperationComparer(Type[] topologicallyOrderedTypes)
+            {
+                _topologicallyOrderedTypes = topologicallyOrderedTypes;
+            }
+
+            public int Compare(IStorageOperation x, IStorageOperation y)
+            {
+                if (x == y)
+                {
+                    return 0;
+                }
+
+                if (x?.DocumentType == null || y?.DocumentType == null)
+                {
+                    return 0;
+                }
+
+                // Maintain order if same document type and same operation
+                if (x.DocumentType == y.DocumentType && x.GetType() == y.GetType())
+                {
+                    return 0;
+                }
+
+                var xIndex = FindIndex(x);
+                var yIndex = FindIndex(y);
+
+                var xIsDelete = x is DeleteWhere || x is DeleteById;
+                var yIsDelete = y is DeleteWhere || y is DeleteById;
+
+                if (xIsDelete != yIsDelete)
+                {
+                    // Arbitrary order if one is a delete but the other is not, because this will force the sorting
+                    // to try and compare these documents against others and fall in to the below checks.
+                    return -1;
+                }
+
+                if (xIsDelete)
+                {
+                    // Both are deletes, so we need reverse topological order to inserts, updates and upserts
+                    return yIndex.CompareTo(xIndex);
+                }
+
+                // Both are inserts, updates or upserts so topological
+                return xIndex.CompareTo(yIndex);
+            }
+
+            private int FindIndex(IStorageOperation x)
+            {
+                // Will loop through up the inheritance chain until reaches the end or the index is found, used
+                // to handle inheritance as topologically sorted array may not have the subclasses listed
+                var documentType = x.DocumentType;
+                var index = 0;
+
+                do
+                {
+                    index = _topologicallyOrderedTypes.IndexOf(documentType);
+                    documentType = documentType.BaseType;
+                } while (index == -1 && documentType != null);
```

Comment:
Are you sure that we won't ever end up in infinite loop?

Reply:
This will break out of the loop once `documentType.BaseType` becomes null, which happens once you hit `object` (goes up the type hierarchy of `documentType`).